### PR TITLE
8276721: G1: Refine G1EvacFailureObjectsSet::iterate

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Currently, G1EvacFailureObjectsSet::iterate(ObjectClosure* closure) is not a const method, but it looks like a const method. It will be a surprise for reader, i.e. you would not expect that an iteration method is destructive, clearing out the set it just iterated over.
We should refine it in some way.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276721](https://bugs.openjdk.java.net/browse/JDK-8276721): G1: Refine G1EvacFailureObjectsSet::iterate

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to 21188299c5e57f101e67a77349f7d69899305d2d
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to 21188299c5e57f101e67a77349f7d69899305d2d

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6303/head:pull/6303` \
`$ git checkout pull/6303`

Update a local copy of the PR: \
`$ git checkout pull/6303` \
`$ git pull https://git.openjdk.java.net/jdk pull/6303/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6303`

View PR using the GUI difftool: \
`$ git pr show -t 6303`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6303.diff">https://git.openjdk.java.net/jdk/pull/6303.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — CHANGES_REQUESTED** [review]
I do not think this change solves what it was intended to do: `G1EvacFailureObjectsSet:handle_iteration()` is still destructive, and the rest of the changes seem to just be moving around code, and making more stuff visible in the header file, overall complicating the code imho.

What about renaming `G1EvacFailureObjectsSet:iterate` to `G1EvacFailureObjectsSet:iterate_and_drop/empty/remove`? Or, as mentioned earlier, have an explicit `drop/remove/emtpy` method in `G1EvacFailureObjectsSet` and call it appropriately?
The former would be my preferred solution because it is then called distinctively enough from a normal `iterate` method, and does not completely mess up current code.

The cause of all this may have been the decision to put `G1EvacFailureObjectsSet`, a data structure only interesting during (young) gc that needs some per-gc handling, into the permanently valid `HeapRegion` data structure - I'll handle this refactoring though as mentioned. 

Let's sync with @albertnetymk about this before continuing.

**Comment by albertnetymk** [comment]
> The former would be my preferred solution

Agree; I prefer not having `iterate` in the name, sth like `process_and_drop/empty/remove`, but either is fine to me.

**Comment by Hamlin-Li** [comment]
I was thinking handle_iteration() will be not that surprising for users as iterate(), but seems it's not.
Sure, let me just rename it as process_and_drop.

**Comment by tschatzl** [comment]
> I was thinking handle_iteration() will be not that surprising for users as iterate(), but seems it's not. Sure, let me just rename it as process_and_drop.

The main issue I have with the current change is that this is a changeset with 160 changed lines for the purpose of renaming two methods which are each called once. That seems completely off.

**Comment by Hamlin-Li** [comment]
Agree, seems it's too much code for such a trivial issue. :)

File: src/hotspot/share/gc/g1/g1EvacFailureObjectsSet.hpp [hunk 1]
```diff
@@ -74,7 +74,7 @@ class G1EvacFailureObjectsSet {
 
   // Apply the given ObjectClosure to all objects that failed evacuation. Objects
   // are passed in increasing address order.
-  void iterate(ObjectClosure* closure);
+  void process_and_drop(ObjectClosure* closure);
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1EvacFailureObjectsSet.hpp:77]: The comment needs some update, something like: "Apply the given ObjectClosure to all objects that failed evacuation and empties the list after processing. Objects are passed in increasing address order."

**Review by tschatzl — APPROVED** [review]
Lgtm apart from the one comment about the comment.

**Review by albertnetymk — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
Thanks @tschatzl @albertnetymk for your reviews.

/integrate

**Pull request closed without merging**